lib/std/simd.zig

```diff
@@ -0,0 +1,395 @@
+
+    try std.testing.expectEqual([4]u32{ 30, 40, 999, 999 }, shiftElementsDown(base, 2, 999));
+    try std.testing.expectEqual([4]u32{ 999, 999, 10, 20 }, shiftElementsUp(base, 2, 999));
+    try std.testing.expectEqual([4]u32{ 20, 30, 40, 10 }, rotateElementsDown(base, 1));
+    try std.testing.expectEqual([4]u32{ 40, 10, 20, 30 }, rotateElementsUp(base, 1));
+    try std.testing.expectEqual([4]u32{ 40, 30, 20, 10 }, reverseOrder(base));
+}
+
+pub fn firstTrue(vec: anytype) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const IndexInt = VectorIndex(@TypeOf(vec));
+
+    if (!@reduce(.Or, vec)) {
+        return null;
+    }
+    const indices = @select(IndexInt, vec, countUp(IndexInt, len), @splat(len, ~@as(IndexInt, 0)));
+    return @reduce(.Min, indices);
+}
+
+pub fn lastTrue(vec: anytype) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const IndexInt = VectorIndex(@TypeOf(vec));
+
+    if (!@reduce(.Or, vec)) {
+        return null;
+    }
+    const indices = @select(IndexInt, vec, countUp(IndexInt, len), @splat(len, @as(IndexInt, 0)));
+    return @reduce(.Max, indices);
+}
+
+pub fn countTrues(vec: anytype) VectorCount(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const CountIntType = VectorCount(@TypeOf(vec));
+
+    const one_if_true = @select(CountIntType, vec, @splat(len, @as(CountIntType, 1)), @splat(len, @as(CountIntType, 0)));
+    return @reduce(.Add, one_if_true);
+}
+
+pub fn firstIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return firstTrue(vec == @splat(len, value));
+}
+
+pub fn lastIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return lastTrue(vec == @splat(len, value));
+}
+
+pub fn countElementsWithValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) VectorCount(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return countTrues(vec == @splat(len, value));
+}
+
+test "vector searching" {
+    const base = Vector(8, u32){ 6, 4, 7, 4, 4, 2, 3, 7 };
+
+    try std.testing.expectEqual(@as(?u3, 1), firstIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, 4), lastIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, null), lastIndexOfValue(base, 99));
+    try std.testing.expectEqual(@as(u4, 3), countElementsWithValue(base, 4));
+}
+
+/// Same as accum, but with a user-provided function that mathematically must be associative.
+/// The function must also do nothing when called with identity as either argument.
+/// Set ErrorType to void if func doesn't return an error union.
+pub fn accumWithFunc(
+    comptime hop: isize,
+    vec: anytype,
+    comptime ErrorType: type,
+    comptime func: fn (@TypeOf(vec), @TypeOf(vec)) if (ErrorType == void) @TypeOf(vec) else ErrorType!@TypeOf(vec),
```

Comment:
I would say that allowing the callback to return an error set here is a little overkill. Did you have a specific use case for this in mind?

Reply:
Mainly functions that can return `error.Overflow`.